<a href="https://colab.research.google.com/github/agarwal-peeush/MachineLearning/blob/master/Python/Learn/2_Getting_Cleaning_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np
import pandas as pd


## Read from text files

In [0]:
# pd.read_csv({filename},sep={sep},encoding={encoding})

## Read from relational database

In [0]:
import pymysql

# create a connection object 'conn'
conn = pymysql.connect(host = "localhost",
                      user="root",
                      passwd="12345",
                      db="information_schema")

# create a cursor object c
c = conn.cursor()

# execute query using c.execute
c.execute("select * from engines;")

# getting the first row of data as a tuple
all_rows = c.fetchall()

# to get only the first row, use c.fetchone() instead


df = pd.DataFrame(list(all_rows), columns=["engine", "support", "comment", 
                                           "transactions", "XA", "savepoints"])
df.head()

## Read from websites - Website scrapping

In [0]:
# 'BeautifulSoup' library is used to parse HTML files 
import requests, bs4

# getting HTML data from the Google play web page
url = "https://play.google.com/store/apps/details?id=com.facebook.orca&hl=en"
req = requests.get(url)

# create a bs4 object
# To avoid warnings, provide "html5lib" explicitly
soup = bs4.BeautifulSoup(req.text, "html5lib")
#print(soup)

# getting all the text inside class = "review-body"
reviews = soup.select('.review-body')
print(type(reviews))
print(len(reviews))
print("\n")

# printing an element of the reviews list
print(reviews[6])

## Read from APIs

In [0]:
import requests, json

# Make the request with the coordinates of San Francisco.
parameters = {"lat": 37.78, "lon": -122.41}
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)

# Get the response data as a python object.  Verify that it's a dictionary.
data = response.json()
print(type(data))
print(data)

print(response.headers)
print(response.headers["content-type"])


# Get the response from the API endpoint.
response = requests.get("http://api.open-notify.org/astros.json")
data = response.json()

# 9 people are currently in space.
print(data["number"])
print(data)

<class 'dict'>
{'message': 'success', 'request': {'altitude': 100, 'datetime': 1545565491, 'latitude': 37.78, 'longitude': -122.41, 'passes': 5}, 'response': [{'duration': 337, 'risetime': 1545569461}, {'duration': 636, 'risetime': 1545575058}, {'duration': 594, 'risetime': 1545580879}, {'duration': 483, 'risetime': 1545586787}, {'duration': 512, 'risetime': 1545592642}]}
{'Server': 'nginx/1.10.3', 'Date': 'Sun, 23 Dec 2018 11:56:58 GMT', 'Content-Type': 'application/json', 'Content-Length': '521', 'Connection': 'keep-alive', 'Via': '1.1 vegur'}
application/json
3
{'people': [{'name': 'Oleg Kononenko', 'craft': 'ISS'}, {'name': 'David Saint-Jacques', 'craft': 'ISS'}, {'name': 'Anne McClain', 'craft': 'ISS'}], 'number': 3, 'message': 'success'}


## Reading data from PDF files

In [0]:
import pyPDF2

# Mounting Google Drive locally

In [0]:
# from google.colab import drive
# drive.mount('/content/gdrive')

# Cleaning datasets

In [0]:
import numpy as np
import pandas as pd

df = pd.read_csv("/content/gdrive/My Drive/Colab Notebooks/melbourne.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode      ...       Bathroom  Car  Landsize  \
0  03-09-2016       2.5    3067.0      ...            1.0  1.0     126.0   
1  03-12-2016       2.5    3067.0      ...            1.0  1.0     202.0   
2  04-02-2016       2.5    3067.0      ...            1.0  0.0     156.0   
3  04-02-2016       2.5    3067.0      ...            2.0  1.0       0.0   
4  04-03-2017       2.5    3067.0      ...            2.0  0.0     134.0   

   BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0           NaN        NaN        Yarra  -37.8014    144.9958   
1           NaN        NaN        Yarra  -37.7996    144.9984   
2          79.0     1900.0        Yarra  -37.8079    144.9934   
3           NaN        NaN        Yarra  -37.8114    145.0116   
4         150.0     1900.0        Yarra  -37.8093    144.9944   

              Regionname Propertycount  
0  Northern Metropolitan        4019.0  
1  Northern Metropolitan        4019.0  
2  Northern Metropolitan        4019.0  
3  Northern Metropolitan        4019.0  
4  Northern Metropolitan        4019.0  

[5 rows x 21 columns]

In [0]:
print(df.shape)
print(df.info())

(23547, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
Suburb           23547 non-null object
Address          23547 non-null object
Rooms            23547 non-null int64
Type             23547 non-null object
Price            18396 non-null float64
Method           23547 non-null object
SellerG          23547 non-null object
Date             23547 non-null object
Distance         23546 non-null float64
Postcode         23546 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
BuildingArea     10018 non-null float64
YearBuilt        11540 non-null float64
CouncilArea      15656 non-null object
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       23546 non-null object
Propertycount    23546 non-null float64
dtypes: float64(12), int64(1), object(8)
memory u

In [0]:
df.isnull()

# summing up the missing values (column-wise)
df.isnull().sum()

# columns having at least one missing value
df.isnull().any()

# above is equivalent to axis=0 (by default, any() operates on columns)
df.isnull().any(axis=0) #axis=0 => column-wise, axis=1 => row-wise

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [0]:
# check if rows which have all values missing
df.isnull().all(axis=1)
df.isnull().all(axis=1).sum() # 0 => There are no rows with all column values missing

# sum of missing values in each row
df.isnull().sum(axis=1)

0        3
1        2
2        0
3        3
4        0
5        2
6        0
7        1
8        2
9        2
10       2
11       0
12       1
13       1
14       0
15       9
16       9
17       2
18       0
19       9
20       1
21       9
22       9
23       2
24       0
25       0
26       7
27       9
28       2
29       2
        ..
23517    2
23518    3
23519    3
23520    2
23521    1
23522    4
23523    3
23524    3
23525    1
23526    2
23527    1
23528    5
23529    4
23530    5
23531    9
23532    1
23533    5
23534    2
23535    3
23536    4
23537    2
23538    1
23539    2
23540    2
23541    1
23542    2
23543    8
23544    4
23545    1
23546    2
Length: 23547, dtype: int64

## Treat missing values


1.   Do nothing if algorithm doesn't complaing about missing values
2.   Delete them
3.   Replace with value such as Mean, median, mode, etc. 



In [0]:
# Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [0]:
# remove columns with more than 30% missing values
df = df.drop('BuildingArea', axis=1)
df = df.drop('YearBuilt', axis=1)
df = df.drop('CouncilArea', axis=1)


In [0]:
 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [0]:
# check rows with more than 5 missing values
df[df.isnull().sum(axis=1) > 5]
len(df[df.isnull().sum(axis=1) > 5].index)
100*(len(df[df.isnull().sum(axis=1) > 5].index)/len(df.index))

18.16791948018856

In [0]:
# remove rows with more than 5 missing values
df = df[df.isnull().sum(axis=1) <= 5]

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

In [0]:
# remove rows with Price null
df = df[~np.isnan(df['Price'])]

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

In [0]:
# check Landsize column and try to impute values for this column
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

Notice that min is 0, max is 433014, the mean is 558, median (50%) is 440. There's a significant variation in the 25th and 75th percentile as well. (176 to 651)

Thus imputing this with mean/median seems quite biased, and so we should remove the NaNs.

In [0]:
# removing NaNs in Landsize
df = df[~np.isnan(df['Landsize'])]

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

There's still marginal data with missing values. Bathroom, Car, Latitude, Longitude. Let's first look at Latitude and Longitude

In [0]:
# rows with missing Latitude, Longitude
df[np.isnan(df['Lattitude'])]

Suburb                Address  Rooms Type      Price  \
2572              Burwood           23 Monica St      3    h   990000.0   
3257         Clifton Hill  3/268 Alexandra Pde E      1    u   363000.0   
4485            Footscray          483 Barkly St      3    t   781000.0   
5170         Hampton East           7 Seafoam St      4    t  1185000.0   
10745  Williamstown North         4/9 Adeline St      1    u   355000.0   
13223           Melbourne   1913/228 Abeckett St      3    u  1175000.0   
14008            Brooklyn          9 Richards Ct      3    h   750000.0   
14132     North Melbourne   13/201 Abbotsford St      2    t   755000.0   
14139      Oakleigh South            4 Druitt St      4    h  1205500.0   
14142      Oakleigh South        298 Warrigal Rd      3    h   799999.0   
14143      Oakleigh South            11 Yarra Ct      5    h  1200000.0   
14976            Essendon        9 Washington St      3    h  1520000.0   
15278              Seddon          60 Station Rd      4    h  1500000.0   
15561             Croydon       3 Silvergrass Ct      3    h   630000.0   
16358          Kensington     201/102 Rankins Rd      2    u   876000.0   
16564          Strathmore          48 Lebanon St      3    h   950000.0   
17616          Kensington     109/102 Rankins Rd      1    u   410000.0   
17620         Keysborough          19 Denmark Rd      5    h  1100000.0   
18141               Lalor           83 Rotino Cr      3    t   463000.0   
18614           Mickleham        17 Primavera Dr      4    h   610000.0   
18793             Wollert      13 Strathalbyn Ch      4    h   631000.0   
18995           Greenvale         40 Frontier Av      3    t   470000.0   

      Method        SellerG        Date  Distance  Postcode  Bedroom2  \
2572      VB      Fletchers  17-09-2016      11.7    3125.0       3.0   
3257       S  hockingstuart  27-06-2016       3.4    3068.0       1.0   
4485       S            Jas  27-11-2016       6.4    3011.0       3.0   
5170       S             RT  28-05-2016      14.5    3188.0       4.0   
10745      S        Sweeney  27-11-2016       8.9    3016.0       1.0   
13223     PI           Icon  29-04-2017       2.8    3000.0       3.0   
14008      S  hockingstuart  20-05-2017      10.9    3012.0       3.0   
14132     PI         Nelson  29-04-2017       2.3    3051.0       2.0   
14139      S       Woodards  22-04-2017      14.7    3167.0       4.0   
14142      S       Woodards  29-04-2017      14.7    3167.0       3.0   
14143     PI         Buxton  29-04-2017      14.7    3167.0       5.0   
14976      S           Brad  03-06-2017       7.5    3040.0       3.0   
15278     SP        Sweeney  03-06-2017       5.1    3011.0       4.0   
15561      S        McGrath  17-06-2017      23.0    3136.0       3.0   
16358      S        Rendina  24-06-2017       3.4    3031.0       2.0   
16564     PI      Considine  24-06-2017       8.2    3041.0       3.0   
17616     PI        Rendina  08-07-2017       3.4    3031.0       1.0   
17620     PI       VicHomes  08-07-2017      25.2    3173.0       5.0   
18141      S            HAR  12-08-2017      16.3    3075.0       3.0   
18614      S            Ray  15-07-2017      20.6    3064.0       4.0   
18793     PI  hockingstuart  15-07-2017      25.5    3750.0       4.0   
18995     SP          Barry  22-07-2017      20.4    3059.0       3.0   

       Bathroom  Car  Landsize  Lattitude  Longtitude  \
2572        2.0  2.0     263.0        NaN         NaN   
3257        1.0  1.0       0.0        NaN         NaN   
4485        2.0  2.0      98.0        NaN         NaN   
5170        3.0  1.0     300.0        NaN         NaN   
10745       1.0  1.0      46.0        NaN         NaN   
13223       3.0  2.0       0.0        NaN         NaN   
14008       1.0  2.0     667.0        NaN         NaN   
14132       1.0  1.0    1537.0        NaN         NaN   
14139       2.0  2.0     553.0        NaN         NaN   
14142       2.0  4.0     590.0        NaN  

In [0]:
# Let's look at the summary stats of Latitude/Longitude columns
df.loc[:,['Lattitude','Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

Since, Std deviation is very small, we can impute these missing values with mean of corresponding columns. 

In [0]:
# imputing Lattitude and Longtitude by mean values
df.loc[np.isnan(df['Lattitude']), ['Lattitude']] = df['Lattitude'].mean()
df.loc[np.isnan(df['Longtitude']), ['Longtitude']] = df['Longtitude'].mean()

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

We have Bathroom, Car have missing values with 0.01% and 0.46%. Let's look at the statistics

In [0]:
df.loc[:, ['Bathroom','Car']].describe()

Bathroom           Car
count  13602.000000  13540.000000
mean       1.534921      1.610414
std        0.691834      0.962244
min        0.000000      0.000000
25%        1.000000      1.000000
50%        1.000000      2.000000
75%        2.000000      2.000000
max        8.000000     10.000000

These 2 are integer type variables and thus have values 0, 1, 2, etc. You cannot impute the NaNs by mean or median. Thus, you need to impute them by the mode - the most common occuring value

In [0]:
# converting to type'category
df['Car']  = df['Car'].astype('category')

# displaying frequencies of each category
df['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [0]:
# the most common occuring is 2, so let's impute NaNs by that. 

#imputing NaNs by 2.0
df.loc[pd.isnull(df['Car']), ['Car']] = 2

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.00
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

In [0]:
# Similar to Bathroom

df['Bathroom'] = df['Bathroom'].astype('category')

df['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [0]:
df.loc[pd.isnull(df['Bathroom']),['Bathroom']] = 1

 # Summing up the missing values (column-wise) %
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64

There are no missing values now. Let's finally look at how many rows (apart from 3 columns) we have lost in the process(originally we had 23547):

In [0]:
df.shape

(13603, 18)

In [0]:
100 * (len(df.index)/23547)

57.769567248481756

Thus we have lost about 42% of observations in cleaning the missing values. 